pkg/loop/telem.go

```diff
@@ -2,34 +2,118 @@ package loop
 
 import (
 	"context"
+	"os"
+	"runtime/debug"
+	"time"
 
 	grpcprom "github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
 	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
 	"go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 
 	"github.com/smartcontractkit/chainlink-relay/pkg/loop/internal"
 )
 
 type GRPCOpts = internal.GRPCOpts
 
-// SetupTelemetry initializes open telemetry and returns GRPCOpts with telemetry interceptors.
-func SetupTelemetry(registerer prometheus.Registerer) GRPCOpts {
-	otel.SetTracerProvider(sdktrace.NewTracerProvider(
-		sdktrace.WithSampler(sdktrace.AlwaysSample()),
-	))
-	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+type TracingConfig struct {
+	// NodeAttributes are the attributes to attach to traces.
+	NodeAttributes map[string]string
+
+	// Enables tracing; requires a collector to be provided
+	Enabled bool
 
+	// Collector is the address of the OTEL collector to send traces to.
+	CollectorTarget string
+
+	// SamplingRatio is the ratio of traces to sample. 1.0 means sample all traces.
+	SamplingRatio float64
+}
+
+// NewGRPCOpts initializes open telemetry and returns GRPCOpts with telemetry interceptors.
+// It is called from the host and each plugin - intended as there is bidirectional communication
+func NewGRPCOpts(registerer prometheus.Registerer) GRPCOpts {
 	if registerer == nil {
 		registerer = prometheus.DefaultRegisterer
 	}
 	return GRPCOpts{DialOpts: dialOptions(registerer), NewServer: newServerFn(registerer)}
 }
 
+// SetupTracing initializes open telemetry with the provided config.
+// It sets the global trace provider and opens a connection to the configured collector.
+func SetupTracing(config TracingConfig) error {
+	if !config.Enabled {
+		return nil
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+	// Enough to shutdown the underlying connection since DialContext is used in blocking mode
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, config.CollectorTarget,
+		// Note the use of insecure transport here. TLS is recommended in production.
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		// safe usage in this case https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#dialing-in-grpc
+		grpc.WithBlock(),
```

Comment:
Can you elaborate on why this is safe? I was expecting the opposite. Our policy is to start fast and without external dependencies which block or may be unavailable. We normally connect (and reconnect as necessary) in the background.

Reply:
Even if it were _safe_, why is it preferable? The default behavior of a deferred internal connection is usually a feature - why give that up?

Reply:
Actually, it looks like you don't have to worry about this detail at all, because the trace exporter can handle it for us, if we pass an endpoint and options instead of a connection:
```go
	// Set up a trace exporter
	// Shutting down the traceExporter will not shutdown the underlying connection.
	traceExporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithEndpoint(config.CollectorTarget), otlptracegrpc.WithDialOption(
		// Note the use of insecure transport here. TLS is recommended in production.
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	))
	if err != nil {
		return err
	}
```

Reply:
> The default behavior of a deferred internal connection is usually a feature - why give that up?

When I wrote this, I didn't want `SetupTelemetry` (now `SetupTracing`) to return before a connection could be made to the otel-collector in cases it could take sufficiently long enough to connect to the collector such that early traces would be dropped. 

That is an assumption though, which is that if the otel-collector isn't properly connected to before traces are generated, the traces will be dropped until it is properly connected to.

Reply:
I don't think it make sense to treat trace collection as critical since it compromises our startup time, and with one blocking attempt we give up in cases where it was temporarily unavailable. Regardless, I would hope that the collection is buffered and not synchronous, so they would only be delayed rather than dropped.